### PR TITLE
adding indel tests to hamming distance method

### PR DESCRIPTION
**Pull request by huddlej**

### CODE REVIEW

File: notebooks/scripts/Helpers.py [hunk 1]
```diff
@@ -98,12 +105,19 @@ def get_hamming_distances(genomes):
...
+            result = np.where(np_genomes_b == "0")
+            np_genomes[result] = 0
+            a = (np_genomes=="0")
+            num_indel = (a&~np.r_[[False],a[:-1]]).sum()
             # Find all mismatches between these two genomes.
             mismatches = genome_arrays[i] != genome_arrays[j]
 
             # Count the number of mismatches where both genomes have valid bases.
-            hamming_distances.append((mismatches & valid_bases[i] & valid_bases[j]).sum())
+            hamming_distances.append(((mismatches & valid_bases[i] & valid_bases[j]).sum()) + num_indel)
```
Comment (huddlej) [inline notebooks/scripts/Helpers.py:120]: We want to parameterize the counting of indels so we can enable/disable that counting per pathogen.

File: notebooks/scripts/hamming_distance_from_fasta.py [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
     parser.add_argument("--alignment", required=True, help="aligned FASTA file of diseases")
     parser.add_argument("--output", required=True, help="path for the csv output")
+    parser.add_argument("--indel-distance", action="store_true", help="add indel distances to genetic distances")
```
Comment (huddlej) [inline notebooks/scripts/hamming_distance_from_fasta.py:22]: I tested this script with the MERS alignment first using the default without indels:

```
python3 notebooks/scripts/hamming_distance_from_fasta.py \
  --alignment mers-nextstrain/results/aligned.fasta \
  --output test.csv
```

This command ran as expected. Then, I ran the command again but with indel distances:

```
python3 notebooks/scripts/hamming_distance_from_fasta.py \
  --alignment mers-nextstrain/results/aligned.fasta \
  --indel-distance \
  --output test_with_indels.csv
```

This command failed with the following error:

```python
Traceback (most recent call last):
  File "notebooks/scripts/hamming_distance_from_fasta.py", line 35, in <module>
    args.indel_distance
  File "/Users/jhuddlesfredhutch.org/projects/cartography/notebooks/scripts/Helpers.py", line 112, in get_hamming_distances
    np_genomes = np.array(list(re.sub("-", "0", genomes[i])))
TypeError: 'odict_values' object is not subscriptable
```

It also looks like the doctests for the hamming distance function aren't passing:

```bash
python3 -m doctest notebooks/scripts/Helpers.py
```

Gives:

```python
**********************************************************************
File "notebooks/scripts/Helpers.py", line 77, in Helpers.get_hamming_distances
Failed example:
    get_hamming_distances(genomes)
Expected:
    [1, 2, 2]
Got:
    [0, 1, 1]
**********************************************************************
File "notebooks/scripts/Helpers.py", line 80, in Helpers.get_hamming_distances
Failed example:
    get_hamming_distances(genomes)
Expected:
    [1, 1, 1]
Got:
    [0, 0, 0]
**********************************************************************
File "notebooks/scripts/Helpers.py", line 83, in Helpers.get_hamming_distances
Failed example:
    get_hamming_distances(genomes)
Expected:
    [3, 2, 4]
Got:
    [2, 0, 2]
**********************************************************************
1 items had failures:
   3 of   8 in Helpers.get_hamming_distances
***Test Failed*** 3 failures.
```

Can you work through the changes needed to get these tests passing before we do more here, @nandsra21?

**Review by huddlej — COMMENTED** [review]
I had a major comment below about getting the doctests to pass before we move forward on this.

**Comment by nandsra21** [comment]
@huddlej I think I've fixed the issues outlined above - can you check and let me know?

**Comment by huddlej** [comment]
@nandsra21 The most recent commits fix the indel counting logic, adding a check that excludes leading and trailing gaps from the total indel-based distances. Can you try updating the MERS workflow to use indel counts and then regenerate the MERS embeddings to see how it works? If it looks ok, we can update the SARS-CoV-2 workflow to use indels, too.

**Review by huddlej — COMMENTED** [review]
@nandsra21 This branch works for me as expected now. You may want to check out the most recent commits where I move indel counting to the embed package, update the Snakemake workflows to use `--indel-distance` once during distance matrix calculation, and remove the duplicate `embed.py` script in favor of our embed package. It looks like including indels in distance calculations only has a small effect on embeddings, but I think this will be a useful addition for the paper and future analyses (especially since phylogenetic methods can't support indels as easily).

**Pull request merged**